core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java

```diff
@@ -0,0 +1,130 @@
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * The {@link LocalPageStore} is an implementation of {@link PageStore} which
+ * stores all pages in a directory somewhere on the local disk.
+ */
+@NotThreadSafe
+public class LocalPageStore implements PageStore {
+
+  private final ResourcePool<ByteBuffer> mBuffers;
+
+  private final String mRoot;
+
+  /**
+   * Creates a new instance of {@link LocalPageStore}.
+   *
+   * @param rootDir the root directory where pages are stored
+   */
+  public LocalPageStore(String rootDir) {
+    this(rootDir, Constants.MB, 32 * Constants.MB);
+  }
+
+  /**
+   * Creates a new instance of {@link LocalPageStore}.
+   *
+   * @param rootDir the root directory where pages are stored
+   * @param bufferSize the size in bytes of the internal buffers used to copy data
+   * @param maxSize the size in bytes that all internal buffers should use
+   */
+  public LocalPageStore(String rootDir, int bufferSize, int maxSize) {
+    mRoot = rootDir;
+    int numBuffers = maxSize / bufferSize;
+    mBuffers = new ResourcePool<ByteBuffer>(numBuffers) {
+      @Override
+      public void close() {
+      }
+
+      @Override
+      protected ByteBuffer createNewResource() {
+        return ByteBuffer.wrap(new byte[bufferSize]);
+      }
+    };
+  }
+
+  @Override
+  public int put(long fileId, long pageIndex, ReadableByteChannel src) throws IOException {
+    Path p = getFilePath(fileId, pageIndex);
+    if (!Files.exists(p)) {
+      Files.createDirectories(p.getParent());
+      Files.createFile(p);
+    }
+    int written = 0;
+    try (FileOutputStream fos = new FileOutputStream(p.toFile(), false)) {
+      ByteBuffer b = mBuffers.acquire();
+      b.clear();
+      try (FileChannel chan = fos.getChannel()) {
+        while (src.read(b) > 0) {
+          b.flip();
+          written += chan.write(b);
+          b.clear();
+        }
+      } finally {
+        mBuffers.release(b);
+      }
+    }
+    return written;
+  }
+
+  @Override
+  public int get(long fileId, long pageIndex, WritableByteChannel dst) throws IOException {
+    int count = 0;
+    Path p = getFilePath(fileId, pageIndex);
+    if (!Files.exists(p)) {
+      throw new FileNotFoundException(p.toString());
+    }
+    try (FileInputStream fis = new FileInputStream(p.toFile())) {
+      ByteBuffer b = mBuffers.acquire();
+      b.clear();
+      try (FileChannel chan = fis.getChannel()) {
+        while (chan.read(b) > 0) {
+          b.flip();
+          count += dst.write(b);
+          b.clear();
+        }
+      } finally {
+        mBuffers.release(b);
+      }
+    }
+    return count;
+  }
+
+  @Override
+  public void delete(long fileId, long pageIndex) throws IOException {
+    Files.delete(getFilePath(fileId, pageIndex));
```

Comment:
do we just delete the file not the parent dirs? will a lot of empty dirs be left after running this a while?

Reply:
updated